### PR TITLE
Propagate types through blocks when they are called or escape.

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Comment by kasperl** [comment]
Fixes #2439.

File: tests/type_propagation/block-local-test.toit [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2024 Toitware ApS.
+// Use of this source code is governed by a Zero-Clause BSD license that can
+// be found in the tests/LICENSE file.
+
+main:
+  test-single
+  test-multi
+  test-outer
+  test-recursive
```
Comment (floitsch) [inline tests/type_propagation/block-local-test.toit:9]: Add test where a function gets two blocks and they influence each other's types.

**Review by floitsch — APPROVED** [review]
LGTM.

File: tests/type_propagation/block-local-test.toit (hunk 1 above)
Comment (kasperl) [reply tests/type_propagation/block-local-test.toit:9]: Done.

File: src/compiler/propagation/type_propagator.cc [hunk 2]
```diff
@@ -354,11 +372,34 @@ void TypePropagator::call_method(MethodTemplate* caller,
                                  TypeScope* scope,
                                  uint8* site,
                                  Method target,
-                                 std::vector<ConcreteType>& arguments) {
+                                 std::vector<ConcreteType>& arguments,
+                                 std::vector<Worklist*>& worklists) {
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc]: Remove the argument again.

File: src/compiler/propagation/type_propagator.cc [hunk 3]
```diff
@@ -354,11 +372,34 @@ void TypePropagator::call_method(MethodTemplate* caller,
...
   TypeStack* stack = scope->top();
   int arity = target.arity();
   int index = arguments.size();
   if (index == arity) {
+    // Collect all block arguments. We cannot handle block arguments
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc]: Fixed comment.

**Pull request merged**